# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent


### 1. iNaturalist statuses

In [161]:
import sys
import os
projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf
import pandas as pd

sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
taxastatus.head(3)

id taxon_id  user_id place_id source_id  \
0  166449    38493  1138587     7830             
1  234788   918383   702203     9994             
2  234789   918383   702203     7308             

                            authority status  \
0  Flora and Fauna Guarantee Act 1988     CR   
1           Atlas of Living Australia    NT    
2           Atlas of Living Australia     LC   

                                                 url description geoprivacy  \
0                                                                  obscured   
1  https://bie.ala.org.au/species/https://id.biod...                          
2  https://bie.ala.org.au/species/https://id.biod...                          

   ...         genus specificEpithet infraspecificEpithet  \
0  ...     Eulamprus      kosciuskoi                        
1  ...  Chiloschista     phyllorhiza                        
2  ...  Chiloschista     phyllorhiza                        

               modified            scientificName taxonRank  \
0  2021-03-01T10:35:01Z      Eulamprus kosciuskoi   species   
1  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   
2  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   

                                          references preferred_common_name  \
0  http://reptile-database.reptarium.cz/search.ph...                         
1  http://www.catalogueoflife.org/annual-checklis...                         
2  http://www.catalogueoflife.org/annual-checklis...                         

  is_active current_synonymous_taxon_ids  
0                                         
1                                         
2                                         

[3 rows x 34 columns]

In [162]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])
inatstatuses = filter_state_statuses("Northern Territory|NT NRETAS", " ")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id user_id place_id source_id                  authority  \
1001    162724  1134239  702203     9994                   Northern Territory   
927     165448    12647  702203     9994                   Northern Territory   
289     263901  1289379  702203     9994            Atlas of Living Australia   
1251    152435    19250  708886     9994     16652                  NT NRETAS   
215     170163    20166  702203     9994            Atlas of Living Australia   
1249    152433    38633  708886     9994     16652                  NT NRETAS   
1252    152436    40743  708886     9994     16652                  NT NRETAS   
1248    152432    41326  708886     9994     16652                  NT NRETAS   
986     139906   698942             9994                                        
1247    152431    73180  708886     9994     16652                  NT NRETAS   
1250    152434    74425  708886     9994     16652                  NT NRETAS   
1       234788   918383  702203     9994            Atlas of Living Australia   

             status                                                url  \
1001             LC  https://bie.ala.org.au/species/https://id.biod...   
927              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
289              VU  https://bie.ala.org.au/species/urn:lsid:biodiv...   
1251     endangered  https://lists.ala.org.au/speciesListItem/list/...   
215              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
1249     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1252     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1248     endangered  https://lists.ala.org.au/speciesListItem/list/...   
986   least concern  https://bie.ala.org.au/species/http://id.biodi...   
1247     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1250     endangered  https://lists.ala.org.au/speciesListItem/list/...   
1               NT   https://bie.ala.org.au/species/https://id.biod...   

                          description geoprivacy  ...         genus  \
1001                                        open  ...    Cryptandra   
927                                         open  ...    Epthianura   
289                                         open  ...      Chloebia   
1251                                    obscured  ...     Polytelis   
215                                               ...         Ninox   
1249                                    obscured  ...   Bellatorias   
1252                                    obscured  ...                 
1248                                    obscured  ...    Macroderma   
986   Atlas of Living Australia (ALA)             ...      Duboisia   
1247                                     private  ...     Pezoporus   
1250                                    obscured  ...  Hipposideros   
1                                                 ...  Chiloschista   

     specificEpithet infraspecificEpithet              modified  \
1001         gemmata                       2020-09-25T18:01:37Z   
927           crocea                       2021-09-17T08:46:17Z   
289         gouldiae                       2022-10-20T02:55:37Z   
1251      alexandrae                       2019-08-27T01:09:01Z   
215        connivens                       2021-07-28T02:17:03Z   
1249           obiri                       2019-04-30T15:19:21Z   
1252                                                              
1248           gigas                       2019-08-27T01:58:05Z   
986        hopwoodii                       2019-02-16T13:19:18Z   
1247    occidentalis                       2019-08-27T01:09:27Z   
1250       inornatus                       2019-01-05T10:25:06Z   
1        phyllorhiza                       2022-01-08T03:30:36Z   

                scientificName taxonRank  \
1001        Cryptandra gemmata   species   
927          Epthianura crocea   species   
289          Chloebia gouldiae   species   
1251      Polyte

### 2. iNaturalist taxonomy

In [163]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists
Sensitive list: `geoprivacy` = `obscured`
overrides

In [164]:
# %%script echo skipping # comment this line to download dataset from lists.ala.org.au the web and save locally
# Download lists data. Retrieve binomial and trinomial names from GBIF. Save locally to CSV

sensitivelist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr492?max=10000&includeKVP=true")
sensitivelist = lf.kvp_to_columns(sensitivelist)
sensitivelist.to_csv(sourcedir + "nt-ala-sensitive.csv", index=False)

conservationlist = lf.download_ala_list("https://lists-test.ala.org.au/ws/speciesListItems/dr651?max=10000&includeKVP=true")
conservationlist = lf.kvp_to_columns(conservationlist)
conservationlist.to_csv(sourcedir + "nt-ala-conservation.csv", index=False)

In [165]:
# Read sensitive list data
sensitivelist = pd.read_csv(sourcedir + "nt-ala-sensitive.csv", dtype=str)
sensitivelist['status'] = 'sensitive'
sensitivelist['nt_geoprivacy'] = 'obscured'
# sensitivelist = sensitivelist.rename(columns={'scientificName':'nt_scientificName'})
sensitivelist

id                    name               commonName  \
0  2058921        Macroderma gigas                Ghost Bat   
1  2058928   Hipposideros stenotis  Northern Leaf-nosed Bat   
2  2058925   Hipposideros inornata    Arnhem Leaf-nosed Bat   
3  2058920  Pezoporus occidentalis             Night Parrot   
4  2058926    Polytelis alexandrae       Alexandra's Parrot   
5  2058929        Falco hypoleucos              Grey Falcon   
6  2058922       Bellatorias obiri      Arnhem Land Egernia   
7  2058923           Attacus wardi               Atlas Moth   
8  2058924      Ogyris iphis doddi             Dodd’s Azure   
9  2058927      Candalides geminus                      NaN   

                  scientificName  \
0               Macroderma gigas   
1          Hipposideros stenotis   
2         Hipposideros inornatus   
3         Pezoporus occidentalis   
4           Polytelis alexandrae   
5  Falco (Hierofalco) hypoleucos   
6              Bellatorias obiri   
7                  Attacus wardi   
8             Ogyris iphis doddi   
9          Erina geminus geminus   

                                                lsid dataResourceUid  \
0  https://biodiversity.org.au/afd/taxa/63bc796a-...           dr492   
1  https://biodiversity.org.au/afd/taxa/26fe0f53-...           dr492   
2  https://biodiversity.org.au/afd/taxa/5d2dab40-...           dr492   
3  https://biodiversity.org.au/afd/taxa/c630f3b0-...           dr492   
4  https://biodiversity.org.au/afd/taxa/be7a08f5-...           dr492   
5  https://biodiversity.org.au/afd/taxa/4c73a934-...           dr492   
6  https://biodiversity.org.au/afd/taxa/2afc8501-...           dr492   
7  https://biodiversity.org.au/afd/taxa/8a05008e-...           dr492   
8  https://biodiversity.org.au/afd/taxa/ae3ab4c9-...           dr492   
9  https://biodiversity.org.au/afd/taxa/14d46baa-...           dr492   

                                           kvpValues          vernacular name  \
0  [{'key': 'vernacular name', 'value': 'Ghost Ba...                Ghost Bat   
1  [{'key': 'vernacular name', 'value': 'Northern...  Northern Leaf-nosed bat   
2  [{'key': 'vernacular name', 'value': 'Arnhem L...    Arnhem Leaf-nosed Bat   
3  [{'key': 'vernacular name', 'value': 'Night Pa...             Night Parrot   
4  [{'key': 'vernacular name', 'value': 'Princess...          Princess Parrot   
5  [{'key': 'vernacular name', 'value': 'Grey Fal...              Grey Falcon   
6  [{'key': 'vernacular name', 'value': 'Arnhemla...       Arnhemland Egernia   
7  [{'key': 'vernacular name', 'value': 'Atlas Mo...               Atlas Moth   
8  [{'key': 'vernacular name', 'value': 'Dodd’s A...             Dodd’s Azure   
9  [{'key': 'vernacular name', 'value': 'Twin Dus...          Twin Dusky-blue   

                     Criteria  \
0  cave roost related records   
1  cave roost related records   
2  cave roost related records   
3                         NaN   
4             nesting records   
5             nesting records   
6                         NaN   
7                         NaN   
8                         NaN   
9      eastern sub-population   

                                      Denature level generalisation  \
0  Round coordinate value to 0.05 decimal degrees...           10km   
1  Round coordinate value to 0.05 decimal degrees...           10km   
2  Round coordinate value to 0.05 decimal degrees...           10km   
3  Round coordinate value to 0.5 decimal degrees ...          100km   
4  Round coordinate value to 0.1 decimal degrees ...           10km   
5  Round coordinate value to 0.1 decimal degrees ...           10km   
6  Round coordinate value to 0.1 decimal degrees ...           10km   
7  Round coordinate value to 0.1 decimal degrees ...           10km   
8  Round coordinate value to 0.1 decimal degrees ...           10km   
9  Round coordinate value to 0.1 decimal degrees ...           10km   

      status nt_geoprivacy  
0  sensitive      obscured  
1  sensitive      obscured  

In [166]:
# Read conservation list data
conservationlist = pd.read_csv(sourcedir + "nt-ala-conservation.csv", dtype=str)
conservationlist['nt_geoprivacy'] = 'open'
# conservationlist = conservationlist.rename(columns={'scientificName':'nt_scientificName'})

conservationlist

id                           name                        commonName  \
0    2640283             Dasyuroides byrnei                            Kowari   
1    2640455              Phascogale calura             Red-tailed Phascogale   
2    2640350               Pseudomys fieldi                   Shark Bay Mouse   
3    2640368             Dasyurus geoffroii                     Western Quoll   
4    2640338                  Acacia latzii                     Latz's Wattle   
..       ...                            ...                               ...   
199  2640385                Leipoa ocellata                        Malleefowl   
200  2640324            Dasyurus hallucatus              Digul [gogo-yimidir]   
201  2640334           Pedionomus torquatus                   Plains-wanderer   
202  2640315         Vincentrachia desmonda            Saddle Creek Rocksnail   
203  2640390  Amytornis modestus indulkanna  Thick-billed Grasswren (western)   

                                scientificName  \
0                           Dasyuroides byrnei   
1                            Phascogale calura   
2                             Pseudomys fieldi   
3                           Dasyurus geoffroii   
4                                Acacia latzii   
..                                         ...   
199                            Leipoa ocellata   
200                        Dasyurus hallucatus   
201                       Pedionomus torquatus   
202                     Vincentrachia desmonda   
203  Amytornis (Amytornis) modestus indulkanna   

                                                  lsid dataResourceUid  \
0    https://biodiversity.org.au/afd/taxa/c342ff42-...           dr651   
1    https://biodiversity.org.au/afd/taxa/36b436b1-...           dr651   
2    https://biodiversity.org.au/afd/taxa/edcf01fa-...           dr651   
3    https://biodiversity.org.au/afd/taxa/a2260672-...           dr651   
4     https://id.biodiversity.org.au/node/apni/2906346           dr651   
..                                                 ...             ...   
199  https://biodiversity.org.au/afd/taxa/c44c9098-...           dr651   
200  https://biodiversity.org.au/afd/taxa/5d7aeda8-...           dr651   
201  https://biodiversity.org.au/afd/taxa/30b4b2e5-...           dr651   
202  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...           dr651   
203  https://biodiversity.org.au/afd/taxa/1309cbb2-...           dr651   

                                             kvpValues                 status  \
0    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
1    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
2    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
3    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
4    [{'key': 'status', 'value': 'Vulnerable'}, {'k...             Vulnerable   
..                                                 ...                    ...   
199  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   
200  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   
201  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   
202  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   
203  [{'key': 'status', 'value': 'Critically Endang...  Critically Endangered   

                   sourceStatus  taxonRemarks nt_geoprivacy  
0    Vulnerable (extinct in NT)        Mammal          open  
1    Vulnerable (extinct in NT)        Mammal          open  
2    Vulnerable (extinct in NT)        Mammal          open  
3    Vulnerable (extinct in NT)        Mammal          open  
4                    Vulnerable           NaN          open  
..                          ...           ...           ...  
199       Critically Endangered          Bird          open  
200       Critically Endangered        Mammal          open  
201       Critic

In [167]:
# Fix duplicate/conflicts detected
conservationlist.at[59, 'nt_geoprivacy'] = 'obscured'
conservationlist.at[159, 'nt_geoprivacy'] = 'obscured'
conservationlist.at[77, 'nt_geoprivacy'] = 'obscured'
conservationlist.at[169, 'nt_geoprivacy'] = 'obscured'
conservationlist.at[176, 'nt_geoprivacy'] = 'obscured'
conservationlist.at[104, 'nt_geoprivacy'] = 'obscured'


In [168]:
# Sensitive list does not have a status.
# Records need status from Conservation list for those species
slist = sensitivelist['name']
slistloc = conservationlist.loc[conservationlist['name'].isin(slist)]
mlist=  sensitivelist.merge(slistloc[['status','name']],how="left",left_on="name",right_on="name")
mlist

id                    name               commonName  \
0  2058921        Macroderma gigas                Ghost Bat   
1  2058928   Hipposideros stenotis  Northern Leaf-nosed Bat   
2  2058925   Hipposideros inornata    Arnhem Leaf-nosed Bat   
3  2058920  Pezoporus occidentalis             Night Parrot   
4  2058926    Polytelis alexandrae       Alexandra's Parrot   
5  2058929        Falco hypoleucos              Grey Falcon   
6  2058922       Bellatorias obiri      Arnhem Land Egernia   
7  2058923           Attacus wardi               Atlas Moth   
8  2058924      Ogyris iphis doddi             Dodd’s Azure   
9  2058927      Candalides geminus                      NaN   

                  scientificName  \
0               Macroderma gigas   
1          Hipposideros stenotis   
2         Hipposideros inornatus   
3         Pezoporus occidentalis   
4           Polytelis alexandrae   
5  Falco (Hierofalco) hypoleucos   
6              Bellatorias obiri   
7                  Attacus wardi   
8             Ogyris iphis doddi   
9          Erina geminus geminus   

                                                lsid dataResourceUid  \
0  https://biodiversity.org.au/afd/taxa/63bc796a-...           dr492   
1  https://biodiversity.org.au/afd/taxa/26fe0f53-...           dr492   
2  https://biodiversity.org.au/afd/taxa/5d2dab40-...           dr492   
3  https://biodiversity.org.au/afd/taxa/c630f3b0-...           dr492   
4  https://biodiversity.org.au/afd/taxa/be7a08f5-...           dr492   
5  https://biodiversity.org.au/afd/taxa/4c73a934-...           dr492   
6  https://biodiversity.org.au/afd/taxa/2afc8501-...           dr492   
7  https://biodiversity.org.au/afd/taxa/8a05008e-...           dr492   
8  https://biodiversity.org.au/afd/taxa/ae3ab4c9-...           dr492   
9  https://biodiversity.org.au/afd/taxa/14d46baa-...           dr492   

                                           kvpValues          vernacular name  \
0  [{'key': 'vernacular name', 'value': 'Ghost Ba...                Ghost Bat   
1  [{'key': 'vernacular name', 'value': 'Northern...  Northern Leaf-nosed bat   
2  [{'key': 'vernacular name', 'value': 'Arnhem L...    Arnhem Leaf-nosed Bat   
3  [{'key': 'vernacular name', 'value': 'Night Pa...             Night Parrot   
4  [{'key': 'vernacular name', 'value': 'Princess...          Princess Parrot   
5  [{'key': 'vernacular name', 'value': 'Grey Fal...              Grey Falcon   
6  [{'key': 'vernacular name', 'value': 'Arnhemla...       Arnhemland Egernia   
7  [{'key': 'vernacular name', 'value': 'Atlas Mo...               Atlas Moth   
8  [{'key': 'vernacular name', 'value': 'Dodd’s A...             Dodd’s Azure   
9  [{'key': 'vernacular name', 'value': 'Twin Dus...          Twin Dusky-blue   

                     Criteria  \
0  cave roost related records   
1  cave roost related records   
2  cave roost related records   
3                         NaN   
4             nesting records   
5             nesting records   
6                         NaN   
7                         NaN   
8                         NaN   
9      eastern sub-population   

                                      Denature level generalisation  \
0  Round coordinate value to 0.05 decimal degrees...           10km   
1  Round coordinate value to 0.05 decimal degrees...           10km   
2  Round coordinate value to 0.05 decimal degrees...           10km   
3  Round coordinate value to 0.5 decimal degrees ...          100km   
4  Round coordinate value to 0.1 decimal degrees ...           10km   
5  Round coordinate value to 0.1 decimal degrees ...           10km   
6  Round coordinate value to 0.1 decimal degrees ...           10km   
7  Round coordinate value to 0.1 decimal degrees ...           10km   
8  Round coordinate value to 0.1 decimal degrees ...           10km   
9  Round coordinate value to 0.1 decimal degrees ...           10km   

    status_x nt_geoprivacy    status_y  
0  sensitive      obscured         NaN  
1  s

In [169]:
sensitivelist['status']  = mlist.status_y.fillna(mlist.status_x)
sensitivelist

id                    name               commonName  \
0  2058921        Macroderma gigas                Ghost Bat   
1  2058928   Hipposideros stenotis  Northern Leaf-nosed Bat   
2  2058925   Hipposideros inornata    Arnhem Leaf-nosed Bat   
3  2058920  Pezoporus occidentalis             Night Parrot   
4  2058926    Polytelis alexandrae       Alexandra's Parrot   
5  2058929        Falco hypoleucos              Grey Falcon   
6  2058922       Bellatorias obiri      Arnhem Land Egernia   
7  2058923           Attacus wardi               Atlas Moth   
8  2058924      Ogyris iphis doddi             Dodd’s Azure   
9  2058927      Candalides geminus                      NaN   

                  scientificName  \
0               Macroderma gigas   
1          Hipposideros stenotis   
2         Hipposideros inornatus   
3         Pezoporus occidentalis   
4           Polytelis alexandrae   
5  Falco (Hierofalco) hypoleucos   
6              Bellatorias obiri   
7                  Attacus wardi   
8             Ogyris iphis doddi   
9          Erina geminus geminus   

                                                lsid dataResourceUid  \
0  https://biodiversity.org.au/afd/taxa/63bc796a-...           dr492   
1  https://biodiversity.org.au/afd/taxa/26fe0f53-...           dr492   
2  https://biodiversity.org.au/afd/taxa/5d2dab40-...           dr492   
3  https://biodiversity.org.au/afd/taxa/c630f3b0-...           dr492   
4  https://biodiversity.org.au/afd/taxa/be7a08f5-...           dr492   
5  https://biodiversity.org.au/afd/taxa/4c73a934-...           dr492   
6  https://biodiversity.org.au/afd/taxa/2afc8501-...           dr492   
7  https://biodiversity.org.au/afd/taxa/8a05008e-...           dr492   
8  https://biodiversity.org.au/afd/taxa/ae3ab4c9-...           dr492   
9  https://biodiversity.org.au/afd/taxa/14d46baa-...           dr492   

                                           kvpValues          vernacular name  \
0  [{'key': 'vernacular name', 'value': 'Ghost Ba...                Ghost Bat   
1  [{'key': 'vernacular name', 'value': 'Northern...  Northern Leaf-nosed bat   
2  [{'key': 'vernacular name', 'value': 'Arnhem L...    Arnhem Leaf-nosed Bat   
3  [{'key': 'vernacular name', 'value': 'Night Pa...             Night Parrot   
4  [{'key': 'vernacular name', 'value': 'Princess...          Princess Parrot   
5  [{'key': 'vernacular name', 'value': 'Grey Fal...              Grey Falcon   
6  [{'key': 'vernacular name', 'value': 'Arnhemla...       Arnhemland Egernia   
7  [{'key': 'vernacular name', 'value': 'Atlas Mo...               Atlas Moth   
8  [{'key': 'vernacular name', 'value': 'Dodd’s A...             Dodd’s Azure   
9  [{'key': 'vernacular name', 'value': 'Twin Dus...          Twin Dusky-blue   

                     Criteria  \
0  cave roost related records   
1  cave roost related records   
2  cave roost related records   
3                         NaN   
4             nesting records   
5             nesting records   
6                         NaN   
7                         NaN   
8                         NaN   
9      eastern sub-population   

                                      Denature level generalisation  \
0  Round coordinate value to 0.05 decimal degrees...           10km   
1  Round coordinate value to 0.05 decimal degrees...           10km   
2  Round coordinate value to 0.05 decimal degrees...           10km   
3  Round coordinate value to 0.5 decimal degrees ...          100km   
4  Round coordinate value to 0.1 decimal degrees ...           10km   
5  Round coordinate value to 0.1 decimal degrees ...           10km   
6  Round coordinate value to 0.1 decimal degrees ...           10km   
7  Round coordinate value to 0.1 decimal degrees ...           10km   
8  Round coordinate value to 0.1 decimal degrees ...           10km   
9  Round coordinate value to 0.1 decimal degrees ...           10km   

       status nt_geoprivacy  
0   sensitive      obscured  
1   sensitive      obscure

In [170]:
slistloc

id                    name                commonName  \
59   2640339           Attacus wardi                Atlas Moth   
77   2640439        Falco hypoleucos               Grey Falcon   
104  2640349    Polytelis alexandrae        Alexandra's Parrot   
159  2640427       Bellatorias obiri  Arnhem Land Gorges Skink   
169  2640464      Ogyris iphis doddi              Dodd’s Azure   
176  2640414  Pezoporus occidentalis              Night Parrot   

                    scientificName  \
59                   Attacus wardi   
77   Falco (Hierofalco) hypoleucos   
104           Polytelis alexandrae   
159              Bellatorias obiri   
169             Ogyris iphis doddi   
176         Pezoporus occidentalis   

                                                  lsid dataResourceUid  \
59   https://biodiversity.org.au/afd/taxa/8a05008e-...           dr651   
77   https://biodiversity.org.au/afd/taxa/4c73a934-...           dr651   
104  https://biodiversity.org.au/afd/taxa/be7a08f5-...           dr651   
159  https://biodiversity.org.au/afd/taxa/2afc8501-...           dr651   
169  https://biodiversity.org.au/afd/taxa/ae3ab4c9-...           dr651   
176  https://biodiversity.org.au/afd/taxa/c630f3b0-...           dr651   

                                             kvpValues      status  \
59   [{'key': 'status', 'value': 'Vulnerable'}, {'k...  Vulnerable   
77   [{'key': 'status', 'value': 'Vulnerable'}, {'k...  Vulnerable   
104  [{'key': 'status', 'value': 'Vulnerable'}, {'k...  Vulnerable   
159  [{'key': 'status', 'value': 'Endangered'}, {'k...  Endangered   
169  [{'key': 'status', 'value': 'Endangered'}, {'k...  Endangered   
176  [{'key': 'status', 'value': 'Endangered'}, {'k...  Endangered   

    sourceStatus  taxonRemarks nt_geoprivacy  
59    Vulnerable  Invertebrate      obscured  
77    Vulnerable          Bird      obscured  
104   Vulnerable          Bird      obscured  
159   Endangered       Reptile      obscured  
169   Endangered  Invertebrate      obscured  
176   Endangered          Bird      obscured

In [171]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist = pd.concat([sensitivelist[['name','status', 'nt_geoprivacy', 'lsid']],
                    conservationlist[['name','status', 'nt_geoprivacy', 'lsid']]]).drop_duplicates()
# retrieve binomial and trinomial names from GBIF
parsednames = lf.gbifparse(statelist)
parsednames.to_csv(sourcedir + "nt-gbif.csv", index=False)
statelist = statelist.merge(parsednames[['scientificName','canonicalName','canonicalNameComplete','type','rankMarker']],how="left",left_on="name",right_on="scientificName")
numfullstatelist = len(statelist.index)
statelist['scientificName'] = statelist['canonicalName']
statelist

name                 status nt_geoprivacy  \
0                 Macroderma gigas              sensitive      obscured   
1            Hipposideros stenotis              sensitive      obscured   
2            Hipposideros inornata              sensitive      obscured   
3           Pezoporus occidentalis             Endangered      obscured   
4             Polytelis alexandrae             Vulnerable      obscured   
..                             ...                    ...           ...   
203                Leipoa ocellata  Critically Endangered          open   
204            Dasyurus hallucatus  Critically Endangered          open   
205           Pedionomus torquatus  Critically Endangered          open   
206         Vincentrachia desmonda  Critically Endangered          open   
207  Amytornis modestus indulkanna  Critically Endangered          open   

                                                  lsid  \
0    https://biodiversity.org.au/afd/taxa/63bc796a-...   
1    https://biodiversity.org.au/afd/taxa/26fe0f53-...   
2    https://biodiversity.org.au/afd/taxa/5d2dab40-...   
3    https://biodiversity.org.au/afd/taxa/c630f3b0-...   
4    https://biodiversity.org.au/afd/taxa/be7a08f5-...   
..                                                 ...   
203  https://biodiversity.org.au/afd/taxa/c44c9098-...   
204  https://biodiversity.org.au/afd/taxa/5d7aeda8-...   
205  https://biodiversity.org.au/afd/taxa/30b4b2e5-...   
206  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...   
207  https://biodiversity.org.au/afd/taxa/1309cbb2-...   

                    scientificName                  canonicalName  \
0                 Macroderma gigas               Macroderma gigas   
1            Hipposideros stenotis          Hipposideros stenotis   
2            Hipposideros inornata          Hipposideros inornata   
3           Pezoporus occidentalis         Pezoporus occidentalis   
4             Polytelis alexandrae           Polytelis alexandrae   
..                             ...                            ...   
203                Leipoa ocellata                Leipoa ocellata   
204            Dasyurus hallucatus            Dasyurus hallucatus   
205           Pedionomus torquatus           Pedionomus torquatus   
206         Vincentrachia desmonda         Vincentrachia desmonda   
207  Amytornis modestus indulkanna  Amytornis modestus indulkanna   

             canonicalNameComplete        type rankMarker  
0                 Macroderma gigas  SCIENTIFIC        sp.  
1            Hipposideros stenotis  SCIENTIFIC        sp.  
2            Hipposideros inornata  SCIENTIFIC        sp.  
3           Pezoporus occidentalis  SCIENTIFIC        sp.  
4             Polytelis alexandrae  SCIENTIFIC        sp.  
..                             ...         ...        ...  
203                Leipoa ocellata  SCIENTIFIC        sp.  
204            Dasyurus hallucatus  SCIENTIFIC        sp.  
205           Pedionomus torquatus  SCIENTIFIC        sp.  
206         Vincentrachia desmonda  SCIENTIFIC        sp.  
207  Amytornis modestus indulkanna  SCIENTIFIC   infrasp.  

[208 rows x 9 columns]

In [172]:
# Identify records that won't comply with iNaturalist species names
noncomply = statelist[statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID']) ]
noncomply

name      status nt_geoprivacy  \
37        Gleichenia sp. Victoria River  Vulnerable          open   
40           Hibbertia sp. South Magela  Vulnerable          open   
60              Triodia sp. Matt Wilson  Vulnerable          open   
62               Typhonium sp. Sandover  Vulnerable          open   
145       Burmannia sp. Bathurst Island  Endangered          open   
147  Erythroxylum sp. Cholmondely Creek  Endangered          open   
151      Livistona mariae subsp. Mariae  Endangered          open   

                                                 lsid  \
37                                       ALA_DR651_28   
40                                       ALA_DR651_31   
60                                       ALA_DR651_51   
62                                       ALA_DR651_53   
145                                     ALA_DR651_139   
147                                     ALA_DR651_141   
151  https://id.biodiversity.org.au/node/apni/8570955   

                  scientificName                canonicalName  \
37        Gleichenia sp.Victoria       Gleichenia sp.Victoria   
40            Hibbertia sp.South           Hibbertia sp.South   
60               Triodia sp.Matt              Triodia sp.Matt   
62         Typhonium sp.Sandover        Typhonium sp.Sandover   
145        Burmannia sp.Bathurst        Burmannia sp.Bathurst   
147  Erythroxylum sp.Cholmondely  Erythroxylum sp.Cholmondely   
151      Livistona mariae subsp.      Livistona mariae subsp.   

           canonicalNameComplete      type rankMarker  
37        Gleichenia sp.Victoria  INFORMAL        sp.  
40            Hibbertia sp.South  INFORMAL        sp.  
60               Triodia sp.Matt  INFORMAL        sp.  
62         Typhonium sp.Sandover  INFORMAL        sp.  
145        Burmannia sp.Bathurst  INFORMAL        sp.  
147  Erythroxylum sp.Cholmondely  INFORMAL        sp.  
151      Livistona mariae subsp.  INFORMAL     subsp.

In [173]:
# remove records that do not comply
statelist = statelist[~statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID']) ]
statelist = pd.DataFrame(statelist[['scientificName','status','nt_geoprivacy','lsid']]).drop_duplicates()
statelist

scientificName                 status nt_geoprivacy  \
0                 Macroderma gigas              sensitive      obscured   
1            Hipposideros stenotis              sensitive      obscured   
2            Hipposideros inornata              sensitive      obscured   
3           Pezoporus occidentalis             Endangered      obscured   
4             Polytelis alexandrae             Vulnerable      obscured   
..                             ...                    ...           ...   
203                Leipoa ocellata  Critically Endangered          open   
204            Dasyurus hallucatus  Critically Endangered          open   
205           Pedionomus torquatus  Critically Endangered          open   
206         Vincentrachia desmonda  Critically Endangered          open   
207  Amytornis modestus indulkanna  Critically Endangered          open   

                                                  lsid  
0    https://biodiversity.org.au/afd/taxa/63bc796a-...  
1    https://biodiversity.org.au/afd/taxa/26fe0f53-...  
2    https://biodiversity.org.au/afd/taxa/5d2dab40-...  
3    https://biodiversity.org.au/afd/taxa/c630f3b0-...  
4    https://biodiversity.org.au/afd/taxa/be7a08f5-...  
..                                                 ...  
203  https://biodiversity.org.au/afd/taxa/c44c9098-...  
204  https://biodiversity.org.au/afd/taxa/5d7aeda8-...  
205  https://biodiversity.org.au/afd/taxa/30b4b2e5-...  
206  https://biodiversity.org.au/afd/taxa/2ff5b7ab-...  
207  https://biodiversity.org.au/afd/taxa/1309cbb2-...  

[201 rows x 4 columns]

In [174]:
parsednames['type'].unique()

array(['SCIENTIFIC', 'INFORMAL'], dtype=object)

In [175]:
# check for duplicates with conflicting information

In [176]:
dupinformation = statelist.groupby('scientificName').filter(lambda x: len(x) > 1)#.sort('size',ascending=False)
dupinformation
#df.groupby('hash').filter(lambda group: len(group) > 1).sort('size', ascending=False)

Empty DataFrame
Columns: [scientificName, status, nt_geoprivacy, lsid]
Index: []

### 4. Equivalent IUCN statuses

In [177]:
# iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild', 'Extinct'}
statelist.groupby(['status'])['status'].count()

status
Critically Endangered     20
Endangered                49
Extinct                   11
Vulnerable               117
sensitive                  4
Name: status, dtype: int64

In [178]:
iucnStatusMappings = {
    'sensitive':'Vulnerable',
    'critically endangered': 'Critically Endangered',
    'endangered': 'Endangered',
    'extinct': 'Extinct',
    'not listed': 'Vulnerable',
    'vulnerable': 'Vulnerable'
}

### 5. Determine best place ID to use

In [179]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()

place_id  place_name          place_display_name    
9994      Northern Territory  Northern Territory, AU    12
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [180]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['scientificName', 'nt_geoprivacy', 'lsid']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses


scientificName nt_geoprivacy  \
138  Abrodictyum obscurum          open   
182  Acacia equisetifolia          open   
14          Acacia latzii          open   
139          Acacia peuce          open   
15    Acacia praetermissa          open   
..                    ...           ...   
155    Xylopia monosperma          open   
61        Zeuxine oblonga          open   
64          Zyzomys maini          open   
165     Zyzomys palatalis          open   
188  Zyzomys pedunculatus          open   

                                                  lsid status_id taxon_id  \
138   https://id.biodiversity.org.au/node/apni/7402565       NaN      NaN   
182   https://id.biodiversity.org.au/node/apni/2890781       NaN      NaN   
14    https://id.biodiversity.org.au/node/apni/2906346       NaN      NaN   
139   https://id.biodiversity.org.au/node/apni/2906202       NaN      NaN   
15    https://id.biodiversity.org.au/node/apni/2894855       NaN      NaN   
..                                                 ...       ...      ...   
155   https://id.biodiversity.org.au/node/apni/2903202       NaN      NaN   
61   https://id.biodiversity.org.au/taxon/apni/5141...       NaN      NaN   
64   https://biodiversity.org.au/afd/taxa/3f638397-...       NaN      NaN   
165  https://biodiversity.org.au/afd/taxa/54aa72cc-...       NaN      NaN   
188  https://biodiversity.org.au/afd/taxa/33ab483d-...       NaN      NaN   

    user_id description iucn authority status geoprivacy place_id  \
138     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
182     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
14      NaN         NaN  NaN       NaN    NaN        NaN      NaN   
139     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
15      NaN         NaN  NaN       NaN    NaN        NaN      NaN   
..      ...         ...  ...       ...    ...        ...      ...   
155     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
61      NaN         NaN  NaN       NaN    NaN        NaN      NaN   
64      NaN         NaN  NaN       NaN    NaN        NaN      NaN   
165     NaN         NaN  NaN       NaN    NaN        NaN      NaN   
188     NaN         NaN  NaN       NaN    NaN        NaN      NaN   

    place_display_name  
138                NaN  
182                NaN  
14                 NaN  
139                NaN  
15                 NaN  
..                 ...  
155                NaN  
61                 NaN  
64                 NaN  
165                NaN  
188                NaN  

[201 rows x 13 columns]

In [181]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = " Territory Parks and Wildlife Conservation Act 1976"
mergedstatuses['new_description'] = "Listed as Threatened - refer to https://nt.gov.au/environment"
biesearchurl = "https://bie.ala.org.au/species/" # eg + "https://id.biodiversity.org.au/node/apni/2894366"
mergedstatuses['new_url'] =  biesearchurl + mergedstatuses['lsid']
mergedstatuses['new_geoprivacy'] = "obscured"
mergedstatuses['new_place_id'] = '9994'  # Northern Territory
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Threatened')
mergedstatuses

scientificName nt_geoprivacy  \
138  Abrodictyum obscurum          open   
182  Acacia equisetifolia          open   
14          Acacia latzii          open   
139          Acacia peuce          open   
15    Acacia praetermissa          open   
..                    ...           ...   
155    Xylopia monosperma          open   
61        Zeuxine oblonga          open   
64          Zyzomys maini          open   
165     Zyzomys palatalis          open   
188  Zyzomys pedunculatus          open   

                                                  lsid status_id taxon_id  \
138   https://id.biodiversity.org.au/node/apni/7402565       NaN      NaN   
182   https://id.biodiversity.org.au/node/apni/2890781       NaN      NaN   
14    https://id.biodiversity.org.au/node/apni/2906346       NaN      NaN   
139   https://id.biodiversity.org.au/node/apni/2906202       NaN      NaN   
15    https://id.biodiversity.org.au/node/apni/2894855       NaN      NaN   
..                                                 ...       ...      ...   
155   https://id.biodiversity.org.au/node/apni/2903202       NaN      NaN   
61   https://id.biodiversity.org.au/taxon/apni/5141...       NaN      NaN   
64   https://biodiversity.org.au/afd/taxa/3f638397-...       NaN      NaN   
165  https://biodiversity.org.au/afd/taxa/54aa72cc-...       NaN      NaN   
188  https://biodiversity.org.au/afd/taxa/33ab483d-...       NaN      NaN   

    user_id description iucn authority status  ... place_id  \
138     NaN         NaN  NaN       NaN    NaN  ...      NaN   
182     NaN         NaN  NaN       NaN    NaN  ...      NaN   
14      NaN         NaN  NaN       NaN    NaN  ...      NaN   
139     NaN         NaN  NaN       NaN    NaN  ...      NaN   
15      NaN         NaN  NaN       NaN    NaN  ...      NaN   
..      ...         ...  ...       ...    ...  ...      ...   
155     NaN         NaN  NaN       NaN    NaN  ...      NaN   
61      NaN         NaN  NaN       NaN    NaN  ...      NaN   
64      NaN         NaN  NaN       NaN    NaN  ...      NaN   
165     NaN         NaN  NaN       NaN    NaN  ...      NaN   
188     NaN         NaN  NaN       NaN    NaN  ...      NaN   

    place_display_name                                      new_authority  \
138                NaN   Territory Parks and Wildlife Conservation Act...   
182                NaN   Territory Parks and Wildlife Conservation Act...   
14                 NaN   Territory Parks and Wildlife Conservation Act...   
139                NaN   Territory Parks and Wildlife Conservation Act...   
15                 NaN   Territory Parks and Wildlife Conservation Act...   
..                 ...                                                ...   
155                NaN   Territory Parks and Wildlife Conservation Act...   
61                 NaN   Territory Parks and Wildlife Conservation Act...   
64                 NaN   Territory Parks and Wildlife Conservation Act...   
165                NaN   Territory Parks and Wildlife Conservation Act...   
188                NaN   Territory Parks and Wildlife Conservation Act...   

                                       new_description  \
138  Listed as Threatened - refer to https://nt.gov...   
182  Listed as Threatened - refer to https://nt.gov...   
14   Listed as Threatened - refer to https://nt.gov...   
139  Listed as Threatened - refer to https://nt.gov...   
15   Listed as Threatened - refer to https://nt.gov...   
..                                                 ...   
155  Listed as Threatened - refer to https://nt.gov...   
61   Listed as Threatened - refer to https://nt.gov...   
64   Listed as Threatened - refer to https://nt.gov...   
165  Listed as Threatened - refer to https://nt.gov...   
188  Listed as Threatened - refer to https://nt.gov...   

                                               new_url new_geoprivacy  \
138  https://bie.ala.org.au/species/https://id.biod...       obscured   
182  https://bie.ala.org.au/species/https://

## Updates

In [182]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action              taxon_name      id taxon_id      status  \
6   UPDATE       Bellatorias obiri  152433    38633  endangered   
63  UPDATE  Hipposideros inornatus  152434    74425  endangered   
1   UPDATE   Hipposideros stenotis  152436    40743  endangered   
0   UPDATE        Macroderma gigas  152432    41326  endangered   
3   UPDATE  Pezoporus occidentalis  152431    73180  endangered   
4   UPDATE    Polytelis alexandrae  152435    19250  endangered   

   iucn_equivalent                                          authority  \
6       Endangered   Territory Parks and Wildlife Conservation Act...   
63      Endangered   Territory Parks and Wildlife Conservation Act...   
1       Endangered   Territory Parks and Wildlife Conservation Act...   
0       Endangered   Territory Parks and Wildlife Conservation Act...   
3       Endangered   Territory Parks and Wildlife Conservation Act...   
4       Endangered   Territory Parks and Wildlife Conservation Act...   

                                                  url geoprivacy place_id  \
6   https://bie.ala.org.au/species/https://biodive...   obscured     9994   
63  https://bie.ala.org.au/species/https://biodive...   obscured     9994   
1   https://bie.ala.org.au/species/https://biodive...   obscured     9994   
0   https://bie.ala.org.au/species/https://biodive...   obscured     9994   
3   https://bie.ala.org.au/species/https://biodive...   obscured     9994   
4   https://bie.ala.org.au/species/https://biodive...   obscured     9994   

     username                                        description  
6   peggydnew  Listed as Threatened - refer to https://nt.gov...  
63  peggydnew  Listed as Threatened - refer to https://nt.gov...  
1   peggydnew  Listed as Threatened - refer to https://nt.gov...  
0   peggydnew  Listed as Threatened - refer to https://nt.gov...  
3   peggydnew  Listed as Threatened - refer to https://nt.gov...  
4   peggydnew  Listed as Threatened - refer to https://nt.gov...

## No status in iNaturalist via straight scientificName match
The NT records that didn't match up to a status in iNaturalist

In [183]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="scientificName",right_on="scientificName")
noinatstatus

scientificName nt_geoprivacy  \
0    Abrodictyum obscurum          open   
1    Acacia equisetifolia          open   
2           Acacia latzii          open   
3            Acacia peuce          open   
4     Acacia praetermissa          open   
..                    ...           ...   
190    Xylopia monosperma          open   
191       Zeuxine oblonga          open   
192         Zyzomys maini          open   
193     Zyzomys palatalis          open   
194  Zyzomys pedunculatus          open   

                                                  lsid status_id taxon_id  \
0     https://id.biodiversity.org.au/node/apni/7402565       NaN      NaN   
1     https://id.biodiversity.org.au/node/apni/2890781       NaN      NaN   
2     https://id.biodiversity.org.au/node/apni/2906346       NaN      NaN   
3     https://id.biodiversity.org.au/node/apni/2906202       NaN      NaN   
4     https://id.biodiversity.org.au/node/apni/2894855       NaN      NaN   
..                                                 ...       ...      ...   
190   https://id.biodiversity.org.au/node/apni/2903202       NaN      NaN   
191  https://id.biodiversity.org.au/taxon/apni/5141...       NaN      NaN   
192  https://biodiversity.org.au/afd/taxa/3f638397-...       NaN      NaN   
193  https://biodiversity.org.au/afd/taxa/54aa72cc-...       NaN      NaN   
194  https://biodiversity.org.au/afd/taxa/33ab483d-...       NaN      NaN   

    user_id description iucn authority status  ...        phylum  \
0       NaN         NaN  NaN       NaN    NaN  ...           NaN   
1       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
2       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
3       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
4       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
..      ...         ...  ...       ...    ...  ...           ...   
190     NaN         NaN  NaN       NaN    NaN  ...           NaN   
191     NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
192     NaN         NaN  NaN       NaN    NaN  ...      Chordata   
193     NaN         NaN  NaN       NaN    NaN  ...      Chordata   
194     NaN         NaN  NaN       NaN    NaN  ...      Chordata   

             class        order       family    genus specificEpithet  \
0              NaN          NaN          NaN      NaN             NaN   
1    Magnoliopsida      Fabales     Fabaceae   Acacia   equisetifolia   
2    Magnoliopsida      Fabales     Fabaceae   Acacia          latzii   
3    Magnoliopsida      Fabales     Fabaceae   Acacia           peuce   
4    Magnoliopsida      Fabales     Fabaceae   Acacia    praetermissa   
..             ...          ...          ...      ...             ...   
190            NaN          NaN          NaN      NaN             NaN   
191     Liliopsida  Asparagales  Orchidaceae  Zeuxine         oblonga   
192       Mammalia     Rodentia      Muridae  Zyzomys           maini   
193       Mammalia     Rodentia      Muridae  Zyzomys       palatalis   
194       Mammalia     Rodentia      Muridae  Zyzomys    pedunculatus   

    infraspecificEpithet              modified taxonRank  \
0                    NaN                   NaN       NaN   
1                    NaN  2022-04-06T22:05:25Z   species   
2                    NaN  2022-04-07T02:06:43Z   species   
3                    NaN  2022-04-06T23:48:13Z   species   
4                    NaN  2022-04-05T02:10:02Z   species   
..                   ...                   ...       ...   
190                  NaN                   NaN       NaN   
191                  NaN  2021-07-16T02:50:44Z   species   
192                  NaN  2019-08-27T01:49:16Z   species   
193                  NaN  2019-11-22T22:46:28Z   species   
194                  NaN  2019-11-22T22:46:29Z   species   

                                            references  
0                                                  NaN  
1                       https://eol.org/

In [184]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions

scientificName nt_geoprivacy  \
1    Acacia equisetifolia          open   
2           Acacia latzii          open   
3            Acacia peuce          open   
4     Acacia praetermissa          open   
5       Acacia undoolyana          open   
..                    ...           ...   
188      Vidumelon wattii          open   
191       Zeuxine oblonga          open   
192         Zyzomys maini          open   
193     Zyzomys palatalis          open   
194  Zyzomys pedunculatus          open   

                                                  lsid status_id taxon_id  \
1     https://id.biodiversity.org.au/node/apni/2890781       NaN      NaN   
2     https://id.biodiversity.org.au/node/apni/2906346       NaN      NaN   
3     https://id.biodiversity.org.au/node/apni/2906202       NaN      NaN   
4     https://id.biodiversity.org.au/node/apni/2894855       NaN      NaN   
5     https://id.biodiversity.org.au/node/apni/2891259       NaN      NaN   
..                                                 ...       ...      ...   
188  https://biodiversity.org.au/afd/taxa/3aedec6e-...       NaN      NaN   
191  https://id.biodiversity.org.au/taxon/apni/5141...       NaN      NaN   
192  https://biodiversity.org.au/afd/taxa/3f638397-...       NaN      NaN   
193  https://biodiversity.org.au/afd/taxa/54aa72cc-...       NaN      NaN   
194  https://biodiversity.org.au/afd/taxa/33ab483d-...       NaN      NaN   

    user_id description iucn authority status  ...        phylum  \
1       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
2       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
3       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
4       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
5       NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
..      ...         ...  ...       ...    ...  ...           ...   
188     NaN         NaN  NaN       NaN    NaN  ...      Mollusca   
191     NaN         NaN  NaN       NaN    NaN  ...  Tracheophyta   
192     NaN         NaN  NaN       NaN    NaN  ...      Chordata   
193     NaN         NaN  NaN       NaN    NaN  ...      Chordata   
194     NaN         NaN  NaN       NaN    NaN  ...      Chordata   

             class            order       family      genus specificEpithet  \
1    Magnoliopsida          Fabales     Fabaceae     Acacia   equisetifolia   
2    Magnoliopsida          Fabales     Fabaceae     Acacia          latzii   
3    Magnoliopsida          Fabales     Fabaceae     Acacia           peuce   
4    Magnoliopsida          Fabales     Fabaceae     Acacia    praetermissa   
5    Magnoliopsida          Fabales     Fabaceae     Acacia      undoolyana   
..             ...              ...          ...        ...             ...   
188     Gastropoda  Stylommatophora   Camaenidae  Vidumelon          wattii   
191     Liliopsida      Asparagales  Orchidaceae    Zeuxine         oblonga   
192       Mammalia         Rodentia      Muridae    Zyzomys           maini   
193       Mammalia         Rodentia      Muridae    Zyzomys       palatalis   
194       Mammalia         Rodentia      Muridae    Zyzomys    pedunculatus   

    infraspecificEpithet              modified taxonRank  \
1                    NaN  2022-04-06T22:05:25Z   species   
2                    NaN  2022-04-07T02:06:43Z   species   
3                    NaN  2022-04-06T23:48:13Z   species   
4                    NaN  2022-04-05T02:10:02Z   species   
5                    NaN  2022-04-05T02:34:14Z   species   
..                   ...                   ...       ...   
188                  NaN  2021-10-29T15:37:54Z   species   
191                  NaN  2021-07-16T02:50:44Z   species   
192                  NaN  2019-08-27T01:49:16Z   species   
193                  NaN  2019-11-22T22:46:28Z   species   
194                  NaN  2019-11-22T22:46:29Z   species   

                                            references  
1                       h

In [185]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'scientificName':'taxon_name',
                                      'id':'taxon_id',
                                      'status_id':'id'})
additions

action            taxon_name   id taxon_id      status iucn_equivalent  \
1      ADD  Acacia equisetifolia  NaN  1253756  Threatened      Vulnerable   
2      ADD         Acacia latzii  NaN  1254327  Threatened      Vulnerable   
3      ADD          Acacia peuce  NaN   465191  Threatened      Vulnerable   
4      ADD   Acacia praetermissa  NaN  1254561  Threatened      Vulnerable   
5      ADD     Acacia undoolyana  NaN  1254884  Threatened      Vulnerable   
..     ...                   ...  ...      ...         ...             ...   
188    ADD      Vidumelon wattii  NaN   114966  Threatened      Vulnerable   
191    ADD       Zeuxine oblonga  NaN   369267  Threatened      Vulnerable   
192    ADD         Zyzomys maini  NaN    45377  Threatened      Vulnerable   
193    ADD     Zyzomys palatalis  NaN    75238  Threatened      Vulnerable   
194    ADD  Zyzomys pedunculatus  NaN    45374  Threatened      Vulnerable   

                                             authority  \
1     Territory Parks and Wildlife Conservation Act...   
2     Territory Parks and Wildlife Conservation Act...   
3     Territory Parks and Wildlife Conservation Act...   
4     Territory Parks and Wildlife Conservation Act...   
5     Territory Parks and Wildlife Conservation Act...   
..                                                 ...   
188   Territory Parks and Wildlife Conservation Act...   
191   Territory Parks and Wildlife Conservation Act...   
192   Territory Parks and Wildlife Conservation Act...   
193   Territory Parks and Wildlife Conservation Act...   
194   Territory Parks and Wildlife Conservation Act...   

                                                   url geoprivacy place_id  \
1    https://bie.ala.org.au/species/https://id.biod...   obscured     9994   
2    https://bie.ala.org.au/species/https://id.biod...   obscured     9994   
3    https://bie.ala.org.au/species/https://id.biod...   obscured     9994   
4    https://bie.ala.org.au/species/https://id.biod...   obscured     9994   
5    https://bie.ala.org.au/species/https://id.biod...   obscured     9994   
..                                                 ...        ...      ...   
188  https://bie.ala.org.au/species/https://biodive...   obscured     9994   
191  https://bie.ala.org.au/species/https://id.biod...   obscured     9994   
192  https://bie.ala.org.au/species/https://biodive...   obscured     9994   
193  https://bie.ala.org.au/species/https://biodive...   obscured     9994   
194  https://bie.ala.org.au/species/https://biodive...   obscured     9994   

      username                                        description  
1    peggydnew  Listed as Threatened - refer to https://nt.gov...  
2    peggydnew  Listed as Threatened - refer to https://nt.gov...  
3    peggydnew  Listed as Threatened - refer to https://nt.gov...  
4    peggydnew  Listed as Threatened - refer to https://nt.gov...  
5    peggydnew  Listed as Threatened - refer to https://nt.gov...  
..         ...                                                ...  
188  peggydnew  Listed as Threatened - refer to https://nt.gov...  
191  peggydnew  Listed as Threatened - refer to https://nt.gov...  
192  peggydnew  Listed as Threatened - refer to https://nt.gov...  
193  peggydnew  Listed as Threatened - refer to https://nt.gov...  
194  peggydnew  Listed as Threatened - refer to https://nt.gov...  

[141 rows x 12 columns]

In [186]:
# write these to the file
pd.concat([updates,additions]).to_csv(sourcedir + "nt.csv", index=False)

In [187]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

scientificName nt_geoprivacy  \
0                 Abrodictyum obscurum          open   
19                  Babingtonia behrii          open   
20                 Baumea arthrophylla          open   
22             Bettongia lesueur graii          open   
32                  Candalides geminus      obscured   
46          Ctenotus rimacola camptris          open   
54                      Dienia montana          open   
55               Dirutrachia sublevata          open   
57                  Elaeocarpus miegei          open   
58                Eleocharis papillosa          open   
59                Endiandra limnophila          open   
63                  Erythrura gouldiae          open   
73             Granulomelon arcigerens          open   
74                Granulomelon gilleni          open   
78               Helicteres macrothrix          open   
79                 Hibbertia brennanii          open   
80                  Hibbertia pancerea          open   
81                 Hibbertia tricornis          open   
82                  Hibiscus brennanii          open   
83                   Hibiscus cravenii          open   
84               Hipposideros inornata      obscured   
88                    Jacksonia divisa          open   
98            Lithomyrtus linariifolia          open   
100                   Luisia corrugata          open   
107       Mesembriomys gouldii gouldii          open   
108  Mesembriomys gouldii melvillensis          open   
114                 Mitrella tiwiensis          open   
115                 Monochoria hastata          open   
122                 Ogyris iphis doddi      obscured   
123            Olearia macdonnellensis          open   
132        Petrogale concinna concinna          open   
142                   Pseudomys fieldi          open   
148                 Schoenus centralis          open   
151                Semotrachia caupona          open   
152                Semotrachia elleryi          open   
153                 Semotrachia emilia          open   
154                   Semotrachia esau          open   
155                 Semotrachia euzyga          open   
156              Semotrachia filixiana          open   
157             Semotrachia huckittana          open   
158               Semotrachia illarana          open   
159              Semotrachia jessieana          open   
160                Semotrachia jinkana          open   
161                Semotrachia rossana          open   
162           Semotrachia runutjirbana          open   
163            Semotrachia winneckeana          open   
164              Simalia oenpelliensis          open   
167                  Sporobolus latzii          open   
171                    Tateropa aemula          open   
173             Trachiopsis victoriana          open   
177                 Typhonium mirabile          open   
179                  Typhonium taylori          open   
189             Vincentrachia desmonda          open   
190                 Xylopia monosperma          open   

                                                  lsid status_id taxon_id  \
0     https://id.biodiversity.org.au/node/apni/7402565       NaN      NaN   
19   https://id.biodiversity.org.au/taxon/apni/5131...       NaN      NaN   
20    https://id.biodiversity.org.au/node/apni/2890775       NaN      NaN   
22   https://biodiversity.org.au/afd/taxa/7a019451-...       NaN      NaN   
32   https://biodiversity.org.au/afd/taxa/14d46baa-...       NaN      NaN   
46   https://biodiversity.org.au/afd/taxa/ef8aa425-...       NaN      NaN   
54   https://id.biodiversity.org.au/taxon/apni/5140...       NaN      NaN   
55   https://biodiversity.org.au/afd/taxa/75839d82-...       NaN      NaN   
57    https://id.biodiversity.org.au/node/apni/2890315       NaN      NaN   
58    https://id.biodiversity.org.au/node/apni/2901498       NaN      NaN   
59    https://id.biodiversity.org.au/node/apni/2886490       NaN      NaN   
63   https://biodiv

### are there any that need to be removed?
NT sensitive list count: 10
NT conservation list count: 204
NT inat statuses count: 12

updates to inat status: 9
additional inat status: 143
NT statuses we can't find a taxon match for in iNaturalist: 62
total:  (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over:  that may need checking

In [188]:
# inat statuses that aren't in added or updated
notaddedupdated = inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]
notaddedupdated

status_id taxon_id user_id place_id source_id                  authority  \
1001    162724  1134239  702203     9994                   Northern Territory   
927     165448    12647  702203     9994                   Northern Territory   
289     263901  1289379  702203     9994            Atlas of Living Australia   
215     170163    20166  702203     9994            Atlas of Living Australia   
986     139906   698942             9994                                        
1       234788   918383  702203     9994            Atlas of Living Australia   

             status                                                url  \
1001             LC  https://bie.ala.org.au/species/https://id.biod...   
927              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
289              VU  https://bie.ala.org.au/species/urn:lsid:biodiv...   
215              LC  https://bie.ala.org.au/species/urn:lsid:biodiv...   
986   least concern  https://bie.ala.org.au/species/http://id.biodi...   
1               NT   https://bie.ala.org.au/species/https://id.biod...   

                          description geoprivacy  ...         genus  \
1001                                        open  ...    Cryptandra   
927                                         open  ...    Epthianura   
289                                         open  ...      Chloebia   
215                                               ...         Ninox   
986   Atlas of Living Australia (ALA)             ...      Duboisia   
1                                                 ...  Chiloschista   

     specificEpithet infraspecificEpithet              modified  \
1001         gemmata                       2020-09-25T18:01:37Z   
927           crocea                       2021-09-17T08:46:17Z   
289         gouldiae                       2022-10-20T02:55:37Z   
215        connivens                       2021-07-28T02:17:03Z   
986        hopwoodii                       2019-02-16T13:19:18Z   
1        phyllorhiza                       2022-01-08T03:30:36Z   

                scientificName taxonRank  \
1001        Cryptandra gemmata   species   
927          Epthianura crocea   species   
289          Chloebia gouldiae   species   
215            Ninox connivens   species   
986         Duboisia hopwoodii   species   
1     Chiloschista phyllorhiza   species   

                                             references preferred_common_name  \
1001  http://plantsoftheworldonline.org/taxon/urn:ls...                         
927   http://www.birdlife.org/datazone/speciesfactsh...                         
289   https://www.birds.cornell.edu/clementschecklis...                         
215   http://www.birdlife.org/datazone/speciesfactsh...                         
986                             https://www.ala.org.au/                         
1     http://www.catalogueoflife.org/annual-checklis...                         

     is_active current_synonymous_taxon_ids  
1001                                         
927                                          
289                                          
215                                          
986                                          
1                                            

[6 rows x 34 columns]

In [189]:
# Stats
numsensitive = len(sensitivelist.index)
numconservation = len(conservationlist.index)
numupdates  = len(updates.index)
numadditions  = len(additions.index)
numnoinatstatus = len(noinatstatus.index)
numunknownToInat = len(unknownToInat.index)
numnotaddedupdated = len(notaddedupdated.index)
numnoncomply = len(noncomply.index)
numcomply = len(statelist.index)
numdupinfo = len(dupinformation.index)
d = {'Sensitive': [numsensitive],
    'Conservation': [numconservation],
    'Statelist merge': [numfullstatelist],
    'Species iNat Comply' : [numcomply],
    'Species iNat non-Comply': [numnoncomply],
    'Duplicate Information': [numdupinfo],
    'Updates': [numupdates],
    'Additions': [numadditions],
    'Not added updated': [numnotaddedupdated],
    'No Inat Status': [numnoinatstatus],
    'Unknown to Inat': [numunknownToInat]}

statsdf = pd.DataFrame(data=d)
statsdf

Sensitive  Conservation  Statelist merge  Species iNat Comply  \
0         10           204              208                  201   

   Species iNat non-Comply  Duplicate Information  Updates  Additions  \
0                        7                      0        6        141   

   Not added updated  No Inat Status  Unknown to Inat  
0                  6             195               54